folding-schemes/src/folding/nova/decider.rs

```diff
@@ -0,0 +1,496 @@
+    S2_ProvingKey: Clone + CanonicalSerialize + CanonicalDeserialize,
+{
+    pub c1_snark_pp: S1_ProvingKey,
+    pub c1_cs_pp: CS1_ProvingKey,
+    pub c2_snark_pp: S2_ProvingKey,
+    pub c2_cs_pp: CS2_ProvingKey,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, CanonicalSerialize, CanonicalDeserialize)]
+pub struct VerifierParam<C1, CS1_VerifyingKey, S1_VerifyingKey, CS2_VerifyingKey, S2_VerifyingKey>
+where
+    C1: CurveGroup,
+    CS1_VerifyingKey: Clone + CanonicalSerialize + CanonicalDeserialize,
+    S1_VerifyingKey: Clone + CanonicalSerialize + CanonicalDeserialize,
+    CS2_VerifyingKey: Clone + CanonicalSerialize + CanonicalDeserialize,
+    S2_VerifyingKey: Clone + CanonicalSerialize + CanonicalDeserialize,
+{
+    pub pp_hash: C1::ScalarField,
+    pub c1_snark_vp: S1_VerifyingKey,
+    pub c1_cs_vp: CS1_VerifyingKey,
+    pub c2_snark_vp: S2_VerifyingKey,
+    pub c2_cs_vp: CS2_VerifyingKey,
+}
+
+/// Onchain Decider, for ethereum use cases
+#[derive(Clone, Debug)]
+pub struct Decider<C1, GC1, C2, GC2, FC, CS1, CS2, S1, S2, FS> {
+    _c1: PhantomData<C1>,
+    _gc1: PhantomData<GC1>,
+    _c2: PhantomData<C2>,
+    _gc2: PhantomData<GC2>,
+    _fc: PhantomData<FC>,
+    _cs1: PhantomData<CS1>,
+    _cs2: PhantomData<CS2>,
+    _s1: PhantomData<S1>,
+    _s2: PhantomData<S2>,
+    _fs: PhantomData<FS>,
+}
+
+impl<C1, GC1, C2, GC2, FC, CS1, CS2, S1, S2, FS> DeciderTrait<C1, C2, FC, FS>
+    for Decider<C1, GC1, C2, GC2, FC, CS1, CS2, S1, S2, FS>
+where
+    C1: CurveGroup,
+    C2: CurveGroup,
+    GC1: CurveVar<C1, CF2<C1>> + ToConstraintFieldGadget<CF2<C1>>,
+    GC2: CurveVar<C2, CF2<C2>> + ToConstraintFieldGadget<CF2<C2>>,
+    FC: FCircuit<C1::ScalarField>,
+    CS1: CommitmentScheme<
+        C1,
+        ProverChallenge = C1::ScalarField,
+        Challenge = C1::ScalarField,
+        Proof = crate::commitment::kzg::Proof<C1>,
+    >,
+    CS2: CommitmentScheme<
+        C2,
+        ProverChallenge = C2::ScalarField,
+        Challenge = C2::ScalarField,
+        Proof = crate::commitment::kzg::Proof<C2>,
+    >,
+    S1: SNARK<C1::ScalarField>,
+    S2: SNARK<C2::ScalarField>,
+    FS: FoldingScheme<C1, C2, FC>,
+    <C1 as CurveGroup>::BaseField: PrimeField,
+    <C2 as CurveGroup>::BaseField: PrimeField,
+    <C1 as Group>::ScalarField: Absorb,
+    <C2 as Group>::ScalarField: Absorb,
+    C1: CurveGroup<BaseField = C2::ScalarField, ScalarField = C2::BaseField>,
+    for<'b> &'b GC1: GroupOpsBounds<'b, C1, GC1>,
+    for<'b> &'b GC2: GroupOpsBounds<'b, C2, GC2>,
+    // constrain FS into Nova, since this is a Decider specifically for Nova
+    Nova<C1, GC1, C2, GC2, FC, CS1, CS2, false>: From<FS>,
+    crate::folding::nova::ProverParams<C1, C2, CS1, CS2, false>:
+        From<<FS as FoldingScheme<C1, C2, FC>>::ProverParam>,
+    crate::folding::nova::VerifierParams<C1, C2, CS1, CS2, false>:
+        From<<FS as FoldingScheme<C1, C2, FC>>::VerifierParam>,
+{
```

Comment:
DAAAAAAAAAMN!! This is getting out of hand really quick...

Reply:
modularity :tada: 🥲

Reply:
Note though, that the usage is clear thanks to arkworks: [folding-schemes/src/folding/nova/decider.rs#L439](https://github.com/privacy-scaling-explorations/sonobe/blob/c831ab5d30cc5d7d79d548e18c4d2118829862dc/folding-schemes/src/folding/nova/decider.rs#L439)
ie.
```rust
type D = Decider<Projective, GVar, Projective2, GVar2, CubicFCircuit<Fr>,
                 KZG<'static, MNT4>, KZG<'static, MNT6>, Groth16<MNT4>, Groth16<MNT6>, N>;
```

where the developer does not need to care about all the trait bounds, since the arkworks objects for the curves and the commitment schemes trait that we have already satisfy all those trait bounds and it's pretty transparent for the developer.
It's just that we need to restrict the Decider object to satisfy these certain properties and relations between the curves used and also between the schemes used, but the developer does not have to deal with that since it comes given by the already existing objects in arkworks and in the various sonobe packages.

Reply:
I know I know!!!

It's just that it was funny to see that monster of a trait bounds statement ahhaha